libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/backpress/BackPressedComponent.kt

```diff
@@ -0,0 +1,84 @@
+package com.mapbox.navigation.dropin.backpress
+
+import androidx.activity.ComponentActivity
+import androidx.activity.OnBackPressedCallback
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.LifecycleOwner
+import androidx.lifecycle.repeatOnLifecycle
+import com.mapbox.navigation.base.ExperimentalPreviewMapboxNavigationAPI
+import com.mapbox.navigation.core.MapboxNavigation
+import com.mapbox.navigation.dropin.NavigationView
+import com.mapbox.navigation.ui.app.internal.Store
+import com.mapbox.navigation.ui.app.internal.destination.DestinationAction
+import com.mapbox.navigation.ui.app.internal.endNavigation
+import com.mapbox.navigation.ui.app.internal.extension.dispatch
+import com.mapbox.navigation.ui.app.internal.navigation.NavigationState
+import com.mapbox.navigation.ui.app.internal.navigation.NavigationStateAction
+import com.mapbox.navigation.ui.app.internal.routefetch.RoutePreviewAction
+import com.mapbox.navigation.ui.app.internal.routefetch.RoutesAction
+import com.mapbox.navigation.ui.base.lifecycle.UIComponent
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.onCompletion
+import kotlinx.coroutines.launch
+
+/**
+ * Back pressed callback for the [NavigationView].
+ *
+ * On back pressed will update the navigation state.
+ *
+ * (FreeDrive) <- (DestinationPreview) <- (RoutePreview) <- (ActiveNavigation)
+ *             <- (Arrival)
+ */
+@ExperimentalPreviewMapboxNavigationAPI
+internal class BackPressedComponent(
+    componentActivity: ComponentActivity,
+    private val store: Store,
+    private val lifecycleOwner: LifecycleOwner,
+) : UIComponent() {
+
+    private val onBackPressedCallback = object : OnBackPressedCallback(false) {
+
+        override fun handleOnBackPressed() {
+            when (store.state.value.navigation) {
+                NavigationState.FreeDrive -> {
+                    error("The callback should be disabled in Free Drive state")
+                }
+                NavigationState.DestinationPreview -> {
+                    store.dispatch(DestinationAction.SetDestination(null))
+                    store.dispatch(NavigationStateAction.Update(NavigationState.FreeDrive))
+                }
+                NavigationState.RoutePreview -> {
+                    store.dispatch(RoutePreviewAction.Ready(emptyList()))
+                    store.dispatch(NavigationStateAction.Update(NavigationState.DestinationPreview))
+                }
+                NavigationState.ActiveNavigation -> {
+                    store.dispatch(RoutesAction.SetRoutes(emptyList()))
+                    store.dispatch(NavigationStateAction.Update(NavigationState.RoutePreview))
+                }
+                NavigationState.Arrival -> {
+                    store.dispatch(endNavigation())
+                }
+            }
+        }
+    }
+
+    init {
+        componentActivity.onBackPressedDispatcher.addCallback(onBackPressedCallback)
```

Comment:
👍 

Also, instead of adding the callback on init it can be added `onAttached`. This does not need an `init`

Reply:
Right now it seems meaningless because it is attached `attachCreated`. But it is likely that this component will be needed to move to `attachResumed` at some point. And we'll need to update this anyways

Reply:
The problem is that `onAttached(MapboxNavigation)` is called too late. It is even called after `onResume` of the activity that hosts this `NavigationView`. So any `OnBackPressedCallback` developer registers in their activity or fragment will have a lower priority that the one added by `NavigationView`. Actually this is the reason why I didn't implement this as `UIComponent` in the first place.

Reply:
> is called too late. It is even called after `onResume`

This is a potential issue. It would be good if `onAttached` was called when `onCreate` is called. Sounds like the `ViewLifecycleRegistry` will not move to a created state until the lifecycle is resumed. This is also, something that may be addressed.

But that may have reasons to keep it the way it is, so good reason to keep it in the `init`

Can you add a comment to describe some of this? It is likely to be revisited at some point.

```
    // Add the callback before onAttached so downstream developers can register higher priority
    // back press callbacks.
    init {
        componentActivity.onBackPressedDispatcher.addCallback(onBackPressedCallback)
```

Reply:
> Actually this is the reason why I didn't implement this as UIComponent in the first place.

Seems like an issue that would exist before as well. Because we don't have lifecycle hooks for re-attach that are not part of the resume/pause.

By the way, that is the way compose works too. Composables will not be created until they are "resumed"

Reply:
Honestly this is looking like a good reason to support the `NavigationViewListener.onBackPressed` because then we have better control of the behavior.

The `re-attach` issue will likely come up again because we are supporting `NavigationView` being re-attached.

Reply:
Scratch what I was saying, the reattach issue would only happen if we move this component outside on the `attachCreated` lifecycle so the comment should be good enough here.